diff --git a/build_config.rb b/build_config.rb
index 3408f19..a7750d6 100644
--- a/build_config.rb
+++ b/build_config.rb
@@ -33,6 +33,10 @@ MRuby::Build.new do |conf|
   #   cc.compile_options = "%{flags} -MMD -o %{outfile} -c %{infile}"
   # end
 
+  conf.cc do |cc|
+    cc.flags = '-fPIC'
+  end
+
   # mrbc settings
   # conf.mrbc do |mrbc|
   #   mrbc.compile_options = "-g -B%{funcname} -o-" # The -g option is required for line numbers
@@ -57,6 +61,11 @@ MRuby::Build.new do |conf|
   #   archiver.archive_options = 'rs %{outfile} %{objs}'
   # end
 
+  conf.archiver do |archiver|
+    archiver.command = 'gcc'
+    archiver.archive_options = '-shared -o %{outfile} %{objs}'
+  end
+
   # Parser generator settings
   # conf.yacc do |yacc|
   #   yacc.command = ENV['YACC'] || 'bison'
diff --git a/tasks/mruby_build.rake b/tasks/mruby_build.rake
index 8438b6c..59fde43 100644
--- a/tasks/mruby_build.rake
+++ b/tasks/mruby_build.rake
@@ -60,7 +60,7 @@ module MRuby
         if ENV['OS'] == 'Windows_NT'
           @exts = Exts.new('.o', '.exe', '.a')
         else
-          @exts = Exts.new('.o', '', '.a')
+          @exts = Exts.new('.o', '', '.so')
         end
 
         build_dir = build_dir || ENV['MRUBY_BUILD_DIR'] || "#{MRUBY_ROOT}/build"
